improve URL supporting

Amazing feature, thank you! This allows us to use Wasm in npm packages that can be used in Web, Node and Webpack projects.

Did you consider not requiring `./` at the beginning of the relative path? Both

```js
let a = new URL("./wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
let b = new URL("wasm/drand.wasm", "file:///Users/me/projects/drand-client/verifying.js");
```

are the same URL but the second one fails with

> Module not found: Error: Can't resolve 'wasm/drand.wasm' in '/Users/simon/opensource/drand/drand-client'

_Originally posted by @webmaster128 in https://github.com/webpack/webpack/pull/11276#issuecomment-694070263_

--- a/declarations/WebpackOptions.d.ts
+++ b/declarations/WebpackOptions.d.ts
@@ -1317,6 +1317,10 @@ export interface ResolveOptions {
 	 */
 	plugins?: ("..." | ResolvePluginInstance)[];
 	/**
+	 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+	 */
+	preferRelative?: boolean;
+	/**
 	 * Custom resolver.
 	 */
 	resolver?: import("enhanced-resolve").Resolver;

--- a/lib/config/defaults.js
+++ b/lib/config/defaults.js
@@ -966,6 +966,13 @@ const getResolveDefaults = ({ cache, context, targetProperties, mode }) => {
 		byDependency: {
 			wasm: esmDeps(),
 			esm: esmDeps(),
+			url: {
+				preferRelative: true
+			},
+			worker: {
+				...esmDeps(),
+				preferRelative: true
+			},
 			commonjs: cjsDeps(),
 			amd: cjsDeps(),
 			// for backward-compat: loadModule

--- a/lib/dependencies/WorkerDependency.js
+++ b/lib/dependencies/WorkerDependency.js
@@ -47,7 +47,7 @@ class WorkerDependency extends ModuleDependency {
 	}
 
 	get category() {
-		return "esm";
+		return "worker";
 	}
 }
 

--- a/package.json
+++ b/package.json
@@ -14,7 +14,7 @@
     "acorn": "^7.4.0",
     "browserslist": "^4.14.3",
     "chrome-trace-event": "^1.0.2",
-    "enhanced-resolve": "^5.0.0",
+    "enhanced-resolve": "^5.1.0",
     "eslint-scope": "^5.1.0",
     "events": "^3.2.0",
     "glob-to-regexp": "^0.4.1",
@@ -28,7 +28,7 @@
     "tapable": "^2.0.0",
     "terser-webpack-plugin": "^4.1.0",
     "watchpack": "^2.0.0",
-    "webpack-sources": "^2.0.0"
+    "webpack-sources": "^2.0.1"
   },
   "peerDependenciesMeta": {
     "webpack-cli": {

--- a/schemas/WebpackOptions.json
+++ b/schemas/WebpackOptions.json
@@ -2525,6 +2525,10 @@
             ]
           }
         },
+        "preferRelative": {
+          "description": "Prefer to resolve module requests as relative request and fallback to resolving as module.",
+          "type": "boolean"
+        },
         "resolver": {
           "description": "Custom resolver.",
           "tsType": "(import('enhanced-resolve').Resolver)"

--- a/test/Defaults.unittest.js
+++ b/test/Defaults.unittest.js
@@ -450,6 +450,9 @@ describe("Defaults", () => {
 		          "...",
 		        ],
 		      },
+		      "url": Object {
+		        "preferRelative": true,
+		      },
 		      "wasm": Object {
 		        "aliasFields": Array [
 		          "browser",
@@ -469,6 +472,27 @@ describe("Defaults", () => {
 		          "module",
 		          "...",
 		        ],
+		      },
+		      "worker": Object {
+		        "aliasFields": Array [
+		          "browser",
+		        ],
+		        "conditionNames": Array [
+		          "import",
+		          "module",
+		          "...",
+		        ],
+		        "extensions": Array [
+		          ".js",
+		          ".json",
+		          ".wasm",
+		        ],
+		        "mainFields": Array [
+		          "browser",
+		          "module",
+		          "...",
+		        ],
+		        "preferRelative": true,
 		      },
 		    },
 		    "cache": false,
@@ -1035,6 +1059,13 @@ describe("Defaults", () => {
 		@@ ... @@
 		-           "browser",
 		@@ ... @@
+		-         "aliasFields": Array [
+		-           "browser",
+		-         ],
+		+         "aliasFields": Array [],
+		@@ ... @@
+		-           "browser",
+		@@ ... @@
 		-       "browser",
 		+       "node",
 		@@ ... @@
@@ -1162,6 +1193,13 @@ describe("Defaults", () => {
 		@@ ... @@
 		-           "browser",
 		@@ ... @@
+		-         "aliasFields": Array [
+		-           "browser",
+		-         ],
+		+         "aliasFields": Array [],
+		@@ ... @@
+		-           "browser",
+		@@ ... @@
 		-       "browser",
 		+       "node",
 		+       "electron",
@@ -1221,6 +1259,13 @@ describe("Defaults", () => {
 		-     "workerWasmLoading": "fetch",
 		+     "workerChunkLoading": "require",
 		+     "workerWasmLoading": "async-node",
+		@@ ... @@
+		-         "aliasFields": Array [
+		-           "browser",
+		-         ],
+		+         "aliasFields": Array [],
+		@@ ... @@
+		-           "browser",
 		@@ ... @@
 		-         "aliasFields": Array [
 		-           "browser",

--- a/test/configCases/asset-url/target-web2/index.js
+++ b/test/configCases/asset-url/target-web2/index.js
@@ -1,5 +1,11 @@
 it("should handle import.meta.url in URL()", () => {
-	const {href} = new URL("./a.js", import.meta.url);
+	const { href } = new URL("./a.js", import.meta.url);
+
+	expect(href).toBe("https://test.cases/path2/a.js");
+});
+
+it("should handle relative paths in URL()", () => {
+	const { href } = new URL("a.js", import.meta.url);
 
 	expect(href).toBe("https://test.cases/path2/a.js");
 });

--- a/types.d.ts
+++ b/types.d.ts
@@ -7041,6 +7041,7 @@ declare interface ResolveOptionsTypes {
 	fullySpecified: boolean;
 	resolveToContext: boolean;
 	restrictions: Set<string | RegExp>;
+	preferRelative: boolean;
 }
 
 /**
@@ -7141,6 +7142,11 @@ declare interface ResolveOptionsWebpackOptions {
 	 * Plugins for the resolver.
 	 */
 	plugins?: ("..." | ResolvePluginInstance)[];
+
+	/**
+	 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+	 */
+	preferRelative?: boolean;
 
 	/**
 	 * Custom resolver.
@@ -7322,6 +7328,10 @@ declare abstract class ResolverFactory {
 						 */
 						plugins?: ("..." | ResolvePluginInstance)[];
 						/**
+						 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+						 */
+						preferRelative?: boolean;
+						/**
 						 * Custom resolver.
 						 */
 						resolver?: Resolver;
@@ -7434,6 +7444,10 @@ declare abstract class ResolverFactory {
 						 */
 						plugins?: ("..." | ResolvePluginInstance)[];
 						/**
+						 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+						 */
+						preferRelative?: boolean;
+						/**
 						 * Custom resolver.
 						 */
 						resolver?: Resolver;
@@ -7545,6 +7559,10 @@ declare abstract class ResolverFactory {
 			 * Plugins for the resolver.
 			 */
 			plugins?: ("..." | ResolvePluginInstance)[];
+			/**
+			 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+			 */
+			preferRelative?: boolean;
 			/**
 			 * Custom resolver.
 			 */
@@ -9315,6 +9333,11 @@ declare interface UserResolveOptions {
 	 * Use only the sync constiants of the file system calls
 	 */
 	useSyncFileSystemCalls?: boolean;
+
+	/**
+	 * Prefer to resolve module requests as relative requests before falling back to modules
+	 */
+	preferRelative?: boolean;
 }
 declare abstract class VariableInfo {
 	declaredScope: ScopeInfo;
@@ -9763,6 +9786,10 @@ declare interface WithOptions {
 			 * Plugins for the resolver.
 			 */
 			plugins?: ("..." | ResolvePluginInstance)[];
+			/**
+			 * Prefer to resolve module requests as relative request and fallback to resolving as module.
+			 */
+			preferRelative?: boolean;
 			/**
 			 * Custom resolver.
 			 */

